Keyring in database

aweakley opened the pull request:
This moves the keyring into the database to avoid duplication and also keep it up to date.  
It would replace #63

aweakley committed:
Move keyring into database

aweakley committed:
Better method name

aweakley committed:
Prevent Memcached warning

aweakley committed:
Ensure active_for_url_regexes has more than one to work with

aweakley committed:
Better style

aweakley committed:
Optionally use _raw_delete for db KeyringItems for speed on large caches

aweakley committed:
Correct typing syntax

aweakley committed:
Correct typing syntax

aweakley committed:
Additional typing syntax fix

aweakley committed:
Additional typing syntax fix

aweakley committed:
Use latest black version

aweakley committed:
Improve test coverage

aweakley committed:
Update docs to include additional setting

aweakley committed:
Increase URL field length

vsalvino commented on wagtailcache/cache.py:
```diff
@@ -339,25 +342,15 @@ def clear_cache(urls: List[str] = []) -> None:
-        # Save the keyring.
-        _wagcache.set("keyring", keyring)
+    if urls:
+        active_keys = KeyringItem.objects.active_for_url_regexes(
```
If this returns a querystring... could you not instead call `KeyringItem.objects.active_for_url_regexes(urls).delete()` ?

vsalvino commented:
Very nice implementation. Thank you! I left a few comments. Also, another bug fix has been merged in recently (adds a try/catch in `UpdateCacheMiddleware` so you may need to rebase to resolve those conflicts).

vsalvino commented on wagtailcache/models.py:
```diff
@@ -0,0 +1,92 @@
+        self.clear_expired()
+        return item
+
+    def bulk_delete_cache_keys(self, keys: List[str]) -> None:
```
Would prefer if keys is a QuerySet. Because during the delete, we are normally querying keys to delete, flattening them as values list, then passing it in here and re-querying. Double hits to the database. Would prefer to only make one database hit.

aweakley committed:
Ruff formatting after rebase

aweakley committed:
Fix reversion from rebase

aweakley committed:
Requested changes for PR #66

aweakley committed:
Mypy ignore rule

aweakley committed:
Support _raw_delete when clearing the whole cache

jacobcolyvan committed:
Add clear_expired_cache_items cmd

aweakley commented:
I've just reverted this to draft because we're finding that calling clear_expired() during set() is not thread safe.

vsalvino commented:
Interesting - are you using ASGI by chance? Or is this caused by concurrent calls to `clear_expired()`?

aweakley committed:
Add optional jitter to cache timeouts

aweakley committed:
Ensure KeyringItem is validated

- 'testserver' is not a valid netloc so fails Django's validation.
- tests for very long URLs

aweakley committed:
Add comment

aweakley committed:
Documentation for clear expired setting

aweakley committed:
Describe jitter setting

aweakley commented:
It's not ASGI, but there are lots of servers and workers. We were getting many concurrent requests creating calls to `set` and they were each creating a `clear_expired` query on the database. Because that was backing up the pages were not being put into the cache quickly, so that led to even more requests creating calls to both things. h
We were finding a lot of pages were expiring at the same time, I guess on the anniversary of when the cache was last cleared. So I added a way to add some random variation to the timeout which smoothed things out a lot but it didn't fix it. 
In the end the fix turned out to be to run `ANALYZE;` on the database and everything became fast again (for some reason auto vacuum didn't catch it) - so a completely separate issue than all the things I'd thought of up to then. 
So now with the database back to normal I think `clear_expired` on `set` would probably be fine, but I've still kept the option to have it run separately. 

I'll tidy it up and re-submit the PR.

aweakley committed:
URLs with GET parameters can be very long

aweakley committed:
Update Wagtail constraint
